src/proof/PGHelp.cc

```diff
@@ -90,31 +90,37 @@ void ProofGraph::getGraphInfo()
 }
 
 std::vector<clauseid_t> ProofGraph::topolSortingTopDown() const {
-    std::vector<clauseid_t> DFS;
-    std::deque<clauseid_t> q;
-    DFS.reserve(getGraphSize());
-    // Enqueue leaves first
-    q.assign(leaves_ids.begin(),leaves_ids.end());
-    do {
-        clauseid_t id = q.front();
-        q.pop_front();
+    std::vector<clauseid_t> order;
+    std::vector<clauseid_t> unprocessed;
+    std::vector<char> marks (getGraphSize(), 0);
+    order.reserve(getGraphSize());
+    unprocessed.push_back(getRoot()->getId());
+    while (not unprocessed.empty()) {
+        auto id = unprocessed.back();
+        auto & mark = marks.at(id);
+        if (mark == 2) { // already processed
+            unprocessed.pop_back();
+            continue;
+        }
         ProofNode * n = getNode(id);
         assert(n);
-        if (not isSetVisited1(id)) {
-            // Process node if both antecedents visited
-            if ((not n->getAnt1() or isSetVisited1(n->getAnt1()->getId())) and
-                (not n->getAnt2() or isSetVisited1(n->getAnt2()->getId()))) {
-                for (clauseid_t resolvent_id: n->getResolvents()) {
-                    if (getNode(resolvent_id)) q.push_back(resolvent_id);
-                }
-                setVisited1(id);
-                DFS.push_back(id);
-            }
+        if (mark == 1 or n->isLeaf()) { // returning from processing parents or is leaf => mark as done
+            mark = 2;
+            order.push_back(id);
+            unprocessed.pop_back();
+            continue;
         }
+        mark = 1;
```

Comment:
I think your code has different semantics.
Maybe `processParent` is a misleading name, because it only adds unvisited parents to the queue.
A child can only be popped and added to the order once the parents have been processed.
We need postorder traversal, while your suggestion would do preorder traversal.

Reply:
I changed the name of the lambda to `enqueueIfNotVisited` which should hopefully better convey its meaning.
We can close the node and add it to the order only when all nodes above it (node on any path to leaves) have also been closed.

Reply:
> I think your code has different semantics. Maybe `processParent` is a misleading name, because it only adds unvisited parents to the queue. A child can only be popped and added to the order once the parents have been processed. We need postorder traversal, while your suggestion would do preorder traversal.

You're right, I thought about it as if the lambda calls were recursive calls and after them, the whole node along with its antecedents would be completely processed. But the recursion is actually implemented by the stack..